### PR TITLE
Add requested claims to access tokens in cache

### PR DESCRIPTION
**Pull request by logdog16**
## Proposed changes

Adding requested claims to the cache + Access Token to conform with my x-plat MSAL C++ PR: https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/1685

## Type of change

- [X] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [X] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

## Additional information

### CODE REVIEW

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m [hunk 1]
```diff
@@ -314,6 +321,21 @@ - (BOOL)matchesWithRealm:(nullable NSString *)realm
         return YES;
     }
 
+    if ([NSString msidIsStringNilOrBlank:self.requestedClaims] && ![NSString msidIsStringNilOrBlank:requestedClaims])
```
Comment (logdog16) [inline IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:324]: I'd like an extra set of eyes on this, as I think there may be a better/cleaner way to do this that I'm overlooking.

Here's a truth table of what I'm trying to do:

|     | A | B | nil | "" |
|-----|---|---|-----|----|
| A   | X |   |     |    |
| B   |   | X |     |    |
| nil |   |   | X   | X  |
| ""  |   |   | X   | X  |

I was having major problems doing string comparison when one side had a value and the other was a nil using IsEqual or IsEqualToString. Is there a string comparison function that would be better here?

Comment (oldalton) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:324]: Would this work? 
result &= (!self.requestedClaims && !requestedClaims) || [self.requestedClaims isEqualToString:requestedClaims];

Comment (ameyapat) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:324]: Can the 2 conditions be separated out? Makes it less confusing imo : 
`if ([NSString msidIsStringNilOrBlank:self.requestedClaims] && [NSString msidIsStringNilOrBlank:requestedClaims]) {`
`           return YES;`
`}`

`if([self.requestedClaims isEqualToString:requestedClaims]) {`
 `          return YES;`
`}`
`return NO;`

or combine it to : 
`([NSString msidIsStringNilOrBlank:self.requestedClaims] && [NSString msidIsStringNilOrBlank:requestedClaims]) || ([self.requestedClaims isEqualToString:requestedClaims])`

Comment (logdog16) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m]: Thank you both very much for the help. Fixed in latest commit.

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m [hunk 2]
```diff
@@ -316,13 +323,11 @@ - (BOOL)matchesWithRealm:(nullable NSString *)realm
 
     if (clientIDMatchingOptions == MSIDSuperSet)
     {
-        if ((clientId && [self.clientId.msidNormalizedString isEqualToString:clientId.msidNormalizedString])
-            || (familyId && [self.familyId.msidNormalizedString isEqualToString:familyId.msidNormalizedString]))
+        if (!(clientId && [self.clientId.msidNormalizedString isEqualToString:clientId.msidNormalizedString])
```
Comment (oldalton) [inline IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:326]: Why this one got updated? Doesn't seem to be related to the changes in this PR.

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m [hunk 3]
```diff
@@ -337,7 +342,15 @@ - (BOOL)matchesWithRealm:(nullable NSString *)realm
         }
     }
 
-    return YES;
+    if(([NSString msidIsStringNilOrBlank:self.requestedClaims] && [NSString msidIsStringNilOrBlank:requestedClaims]) || ([self.requestedClaims isEqualToString:requestedClaims]))
```
Comment (oldalton) [inline IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:345]: nit: add a space after if

**Review by oldalton — CHANGES_REQUESTED** [review]

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m (hunk 2 above)
Comment (logdog16) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:326]: Previously, this if was an OR that returned YES (exiting out) if either condition is true. Given that I added my new requested_claims comparison check at the end of the matcher, after this statement, I inverted the IF to fail out if both conditions are untrue, and to continue evaluation (down to my new check) if the condition is satisfied. 

If I were to revert this change, I'd either have to duplicate my requested_claims check and put a 2nd copy inside this if statement, or I'd have to move it up in the matcher, and IIRC split it into 2 separate if statements as opposed to the single if I put at the end. 

Happy to change it though if you'd rather I don't touch this :)

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m (hunk 3 above)
Comment (logdog16) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m]: Fixed in latest commit :)

File: IdentityCore/src/cache/key/MSIDDefaultCredentialCacheKey.m [hunk 4]
```diff
@@ -62,7 +64,12 @@ - (NSString *)serviceWithType:(MSIDCredentialType)type
     {
         service  = [NSString stringWithFormat:@"%@|%@", service, appKey];
     }
-    
+
+    if (![NSString msidIsStringNilOrBlank:requestedClaims])
+    {
+        service  = [NSString stringWithFormat:@"%@|%@", service, requestedClaims.msidTokenHash];
```
Comment (antrix1989) [inline IdentityCore/src/cache/key/MSIDDefaultCredentialCacheKey.m:70]: Once we save AT with claims, we won't be able to query it without claims anymore, right?

Comment (logdog16) [reply IdentityCore/src/cache/key/MSIDDefaultCredentialCacheKey.m:70]: Correct, by design. 

MSAL CPP will have a slightly different behavior when claims/capabilities are in use. To be specific, if claims are set, unlike other MSALs we do not bypass cache, mirroring the behavior of WAM, where claims are part of the cache key. Therefore, to error on the side of caution, MSAL CPP will be saving AT's with requested_claims set where appropriate, and using the requested_claims key when querying cache. 

I'm prioritizing interoperability between MSAL CPP and Obj-C ADAL/MSAL here, by explicitly not returning an AT w/ claims set when no claims are requested on the cache lookup, and vice versa. This way, existing MSAL's do not have to make any change (just have requested_claims = nil), and MSAL CPP's AT's with claims will coexist side by side, no chance of overwriting or conflicts.

File: IdentityCore/src/cache/token/MSIDCredentialCacheItem.m (hunk 2 above)
Comment (oldalton) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m:325]: Previously, this if was an OR that returned YES (exiting out) if either condition is true. - I think this is by design :) This is needed for some account removal operations in MSAL code where we want to remove all tokens for the same clientId or familyId if my memory serves me well.

Comment (logdog16) [reply IdentityCore/src/cache/token/MSIDCredentialCacheItem.m]: Fixed in latest commit :)

**Review by oldalton — APPROVED** [review]
